go/runtime/bundle/discovery.go

```diff
@@ -0,0 +1,452 @@
+// for the given runtime.
+func (d *Discovery) Queue(runtimeID common.Namespace, manifestHashes []hash.Hash) {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	// Download bundles only if at least one endpoint is configured.
+	if len(d.globalBaseURLs) == 0 && len(d.runtimeBaseURLs[runtimeID]) == 0 {
+		return
+	}
+
+	// Filter out bundles that have already been fetched.
+	var hashes []hash.Hash
+	for _, hash := range manifestHashes {
+		if d.registry.HasBundle(hash) {
+			continue
+		}
+		hashes = append(hashes, hash)
+	}
+
+	// Update the queue with the new hashes.
+	if len(hashes) == 0 {
+		delete(d.manifestHashes, runtimeID)
+		return
+	}
+	d.manifestHashes[runtimeID] = hashes
+
+	// Trigger immediate discovery or download of new bundles.
+	select {
+	case d.discoverCh <- struct{}{}:
+	default:
+	}
+}
+
+// Download tries to download bundles in the queue.
+func (d *Discovery) Download() {
+	d.mu.RLock()
+	runtimeIDs := slices.Collect(maps.Keys(d.manifestHashes))
+	d.mu.RUnlock()
+
+	for _, runtimeID := range runtimeIDs {
+		d.downloadBundles(runtimeID)
+	}
+}
+
+func (d *Discovery) downloadBundles(runtimeID common.Namespace) {
+	// Try to download queued bundles.
+	d.mu.RLock()
+	hashes := d.manifestHashes[runtimeID]
+	d.mu.RUnlock()
+
+	downloaded := make(map[hash.Hash]struct{})
+	for _, hash := range hashes {
+		if err := d.downloadBundle(runtimeID, hash); err != nil {
+			d.logger.Error("failed to download bundle",
+				"err", err,
+				"runtime_id", runtimeID,
+				"manifest_hash", hash.Hex(),
+			)
+			continue
+		}
+		downloaded[hash] = struct{}{}
+	}
+
+	// Remove downloaded bundles from the queue.
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	var pending []hash.Hash
+	for _, hash := range d.manifestHashes[runtimeID] {
+		if _, ok := downloaded[hash]; ok {
+			continue
+		}
+		pending = append(pending, hash)
+	}
+	if len(pending) == 0 {
+		delete(d.manifestHashes, runtimeID)
+		return
+	}
+	d.manifestHashes[runtimeID] = pending
+}
+
+func (d *Discovery) downloadBundle(runtimeID common.Namespace, manifestHash hash.Hash) error {
+	var errs error
+
+	for _, baseURLs := range [][]string{d.runtimeBaseURLs[runtimeID], d.globalBaseURLs} {
+		for _, baseURL := range baseURLs {
+			if err := d.tryDownloadBundle(runtimeID, manifestHash, baseURL); err != nil {
+				errs = errors.Join(errs, err)
+				continue
+			}
+
+			return nil
+		}
+	}
+
+	return errs
+}
+
+func (d *Discovery) tryDownloadBundle(runtimeID common.Namespace, manifestHash hash.Hash, baseURL string) error {
+	filename := fmt.Sprintf("%s%s", manifestHash.Hex(), FileExtension)
+
+	d.logger.Debug("downloading bundle",
+		"runtime_id", runtimeID,
+		"base_url", baseURL,
+		"filename", filename,
+	)
+
+	url, err := url.JoinPath(baseURL, filename)
+	if err != nil {
+		d.logger.Error("failed to construct URL",
+			"err", err,
+			"base_url", baseURL,
+			"filename", filename,
+		)
+		return fmt.Errorf("failed to construct URL: %w", err)
+	}
+
+	src, err := d.fetchBundle(url)
```

Comment:
So this scheme assumes that all bundles are directly downloadable in the registry by their hash. I would like to avoid the need to copy bundles and instead do it like it was suggested in https://github.com/oasisprotocol/oasis-core/issues/5755#issuecomment-2363506316, namely that there would be an indirection step:
* You construct the URL based on the registry base URL and the manifest hash.
* But instead of containing the entire bundle, this URL only contains metadata on where to find the bundle.

This way we can use existing GitHub releases and just point to those releases in these metadata files instead of needing to copy over the actual bundles.

Reply:
Added indirection step. Now, metadata file containing the URL is downloaded first, and then the bundle.